src/Main.hs

```diff
@@ -32,51 +35,51 @@ options = Options
       <*> optional ( argument str
           ( metavar "CONTRACT"
          <> help "Contract inside of file to analyze" ))
-      <*> optional ( strOption
-          ( long "solc-args"
-         <> help "Optional solidity compiler arguments" ))
-      <*> optional ( option auto
-          ( long "epochs"
-         <> help "Optional number of epochs to run coverage guidance" ))
+      <*> optional ( option str
+          ( long "config"
+         <> help "Echidna config file" ))
 
 opts :: ParserInfo Options
 opts = info (options <**> helper)
   ( fullDesc
   <> progDesc "Fuzzing/property based testing of EVM code"
   <> header "Echidna - Ethereum fuzz testing framework" )
 
-
 main :: IO ()
 main = do
-  (Options f c s n) <- execParser opts
-  (v,a,ts) <- loadSolidity f (pack <$> c) (pack <$> s)
-
-  case n of
-    -- RUN WITHOUT COVERAGE
-    Nothing -> do
-      let prop t = (PropertyName $ show t
-                   , ePropertySeq (flip checkETest t) a v 10
-                   )
-
-      _ <- checkParallel . Group (GroupName f) $ map prop ts
-      return ()
-
-    -- RUN WITH COVERAGE
-    Just epochs -> do
-      tests <- mapM (\t -> fmap (t,) (newMVar [])) ts
-      let prop (cov,t,mvar) = (PropertyName $ show t
-                              , ePropertySeqCoverage cov mvar (flip checkETest t) a v 10
-                              )
-
-      replicateM_ (epochs-1) $ do
-        xs <- forM tests $ \(x,y) -> do
-          cov <- readMVar y
-          lastGen <- getCover cov
-          _ <- swapMVar y []
-          return (lastGen,x,y)
-
-        checkParallel . Group (GroupName f) $ map prop xs
+  -- Read cmd line options and load config
+  (Options file contract configFile) <- execParser opts
+  config <- maybe (pure defaultConfig) parseConfig configFile
+
+  -- Load solidity contract and get VM
+  (v,a,ts) <- runReaderT (loadSolidity file (pack <$> contract)) config
+
+  if (config ^. epochs) <= 0
+    -- Run without coverage
+    then do
+    let prop t = (PropertyName $ show t
+                 , runReader (ePropertySeq (flip checkETest t) a v) config
```

Comment:
why not move this out a bit? I think my favorite way to do this would be to parse the config, then do everything after that in a reader with it accessible